bot/config/bot.py

```diff
@@ -0,0 +1,78 @@
+from dataclasses import dataclass
+
+import asyncpg
+
+
+@dataclass(kw_only=True)
+class BotConfig:
+    """
+    Top-level configuration, meaning values that scope over the entire bot, instead of being local to some guild,
+    channel, user etc.
+    """
+
+    discord_token: str
+
+
+class BotConfigModel:
+    __slots__ = "_pool", "_config"
+    _config: BotConfig | None
+
+    def __init__(self, pool: asyncpg.Pool, /):
+        self._pool = pool
+        self._config = None
+
+    async def get_config(self) -> BotConfig:
+        if self._config is None:
+            self._config = await self._select()
+        return self._config
+
+    async def set_config(self, config: BotConfig, /) -> None:
+        await self._update(config)
+        self._config = config
+
+    async def create_initial_config(self, config: BotConfig, /) -> None:
+        await self._insert(config)
+        self._config = config
+
+    async def _select(self) -> BotConfig:
+        async with self._pool.acquire() as conn:
+            row = await conn.fetchrow(
+                """
+                SELECT discord_token FROM bot.bot_config
+                """
+            )
```

Comment:
I'm not sold on raw SQL via `asyncpg` tbh, I think as we get more and more CRUD going, this will get old fast. Maybe consider `sqlalchemy`, but without the `orm` bits?